mtp_transaction_uploader/patterns.py

```diff
@@ -3,18 +3,30 @@
 CREDIT_REF_PATTERN = re.compile(
     '''
     ^
-    [^a-zA-Z]*                    # skip until first letter
-    ([A-Za-z][0-9]{4}[A-Za-z]{2}) # match the prisoner number
-    \D*                           # skip until next digit
-    ([0-9]{1,2})                  # match 1 or 2 digit day of month
-    \D*                           # skip until next digit
-    ([0-9]{1,2})                  # match 1 or 2 digit month
-    \D*                           # skip until next digit
-    ([0-9]{4}|[0-9]{2})           # match 4 or 2 digit year
-    \D*                           # skip until end
+    (
+        [^A-Z]*                                 # skip until first letter
+        (?P<reference_1>[A-Z][0-9]{4}[A-Z]{2})  # match the prisoner number
+        [^0-9]*                                 # skip until next digit
+        (?P<day_1>[0-9]{1,2})                   # match 1 or 2 digit day of month
+        [^0-9]*                                 # skip until next digit
+        (?P<month_1>[0-9]{1,2})                 # match 1 or 2 digit month
+        [^0-9]*                                 # skip until next digit
+        (?P<year_1>[0-9]{4}|[0-9]{2})           # match 4 or 2 digit year
+        [^0-9]*                                 # skip until end
+    |
+        [^0-9]*                                 # skip until first digit
+        (?P<day_2>[0-9]{1,2})                   # match 1 or 2 digit day of month
+        [^0-9]*                                 # skip until next digit
+        (?P<month_2>[0-9]{1,2})                 # match 1 or 2 digit month
+        [^0-9]*                                 # skip until next digit
+        (?P<year_2>[0-9]{4}|[0-9]{2})           # match 4 or 2 digit year
+        [^A-Z]*                                 # skip until next letter
+        (?P<reference_2>[A-Z][0-9]{4}[A-Z]{2})  # match the prisoner number
+        [^A-Z]*                                 # skip until end
+    )
```

Comment:
Can we pull the components of this out? I feel like having one regex contain the reference pattern twice is likely to end up with us inadvertently fixing bugs in one side but not the other. Maybe have `DATE_PATTERN` and `PRISONER_NUMBER_PATTERN` that you can `.format()` into `CREDIT_REF_PATTERN` to append the `_1`/`_2`? I might also be tempted to break this into two (`CREDIT_REF_DATE_SUFFIX_PATTERN` and `CREDIT_REF_DATE_PREFIX_PATTERN`) and match one or other of them? That would then make it more explicit which patterns are doing what, and why, and makes the regex a little lest nested.